consensus/ibft/signer/signer.go

```diff
@@ -0,0 +1,369 @@
+package signer
+
+import (
+	"errors"
+
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/0xPolygon/polygon-edge/validators"
+)
+
+var (
+	ErrEmptyCommittedSeals        = errors.New("empty committed seals")
+	ErrEmptyParentCommittedSeals  = errors.New("empty parent committed seals")
+	ErrInvalidCommittedSealLength = errors.New("invalid committed seal length")
+	ErrInvalidCommittedSealType   = errors.New("invalid committed seal type")
+	ErrRepeatedCommittedSeal      = errors.New("repeated seal in committed seals")
+	ErrNonValidatorCommittedSeal  = errors.New("found committed seal signed by non validator")
+	ErrNotEnoughCommittedSeals    = errors.New("not enough seals to seal block")
+	ErrSignerMismatch             = errors.New("mismatch address between signer and message sender")
+	ErrValidatorNotFound          = errors.New("validator not found in validator set")
+	ErrInvalidValidators          = errors.New("invalid validators type")
+	ErrInvalidValidator           = errors.New("invalid validator type")
+	ErrInvalidSignature           = errors.New("invalid signature")
+)
+
+// Signer is responsible for signing for blocks and messages in IBFT
+type Signer interface {
+	Type() validators.ValidatorType
+	Address() types.Address
+
+	// IBFT Extra
+	InitIBFTExtra(*types.Header, validators.Validators, Sealer)
+	GetIBFTExtra(header *types.Header) (*IstanbulExtra, error)
+
+	// ProposerSeal
+	WriteProposerSeal(*types.Header) (*types.Header, error)
+	EcrecoverFromHeader(*types.Header) (types.Address, error)
+
+	// CommittedSeal
+	CreateCommittedSeal([]byte) ([]byte, error)
+	VerifyCommittedSeal(validators.Validators, types.Address, []byte, []byte) error
+
+	// CommittedSeals
+	WriteCommittedSeals(*types.Header, map[types.Address][]byte) (*types.Header, error)
+	VerifyCommittedSeals(
+		header *types.Header,
+		validators validators.Validators,
+		quorumSize int,
+	) error
+
+	// ParentCommittedSeals
+	GetParentCommittedSeals(header *types.Header) (Sealer, error)
+	VerifyParentCommittedSeals(
+		parent, header *types.Header,
+		parentValidators validators.Validators,
+		quorum int,
+		mustExist bool,
+	) error
+
+	// IBFTMessage
+	SignIBFTMessage([]byte) ([]byte, error)
+	EcrecoverFromIBFTMessage([]byte, []byte) (types.Address, error)
+
+	// Hash of Header
+	CalculateHeaderHash(*types.Header) (types.Hash, error)
+}
+
+// SignerImpl is an implementation that meets Signer
+type SignerImpl struct {
+	keyManager KeyManager
+}
```

Comment:
What's the point of defining an interface than defining the implementation in the same package as the interface ?

Every client code that imports this package is going to see both.

Reply:
Sorry I don't see what you're trying to say. Signer depends on 2 versions of modules by given mode (ECDSA and BLS).

Reply:
@dbrajovic What's your suggestion?